diff --git a/erts/emulator/drivers/common/efile_drv.c b/erts/emulator/drivers/common/efile_drv.c
index 8aff6c1..bc42a87 100644
--- a/erts/emulator/drivers/common/efile_drv.c
+++ b/erts/emulator/drivers/common/efile_drv.c
@@ -78,6 +78,7 @@
 
 #define FILE_OPT_DELAYED_WRITE 0
 #define FILE_OPT_READ_AHEAD    1
+#define FILE_OPT_DELIMITER     2
 
 /* IPREAD variants */
 
@@ -324,6 +325,7 @@ typedef struct {
     ErlDrvPDL       q_mtx;    /* Mutex for the driver queue, known by the emulator. Also used for
 				 mutual exclusion when accessing field(s) below. */
     size_t          write_buffered;
+    char            delimiter;  /* read_line delimiter (def: '\n') */
 #ifdef USE_VM_PROBES
     int             idnum;      /* Unique ID # for this driver thread/desc */
     char            port_str[DTRACE_TERM_BUF_SIZE];
@@ -470,6 +472,7 @@ struct t_data
 	    size_t        read_size; /* in - out */
 	    size_t        nl_pos; /* out */
 	    short         nl_skip; /* out, 0 or 1 */
+	    char          delimiter;
 #if !ALWAYS_READ_LINE_AHEAD
 	    short         read_ahead; /* in, bool */
 #endif
@@ -805,6 +808,7 @@ file_start(ErlDrvPort port, char* command)
     desc->write_error = 0;
     MUTEX_INIT(desc->q_mtx, port); /* Refc is one, referenced by emulator now */
     desc->write_buffered = 0;
+    desc->delimiter = '\n';        /* default line delimiter is '\n' */
 #ifdef  USE_VM_PROBES
     dtrace_drvport_str(port, desc->port_str);
     get_dt_private(0);           /* throw away return value */
@@ -1223,15 +1227,17 @@ static void invoke_read_line(void *data)
 				&read_size);
 	}
 	if ( (d->result_ok = status)) {
+	    char delim   = d->c.read_line.delimiter;
 	    void *nl_ptr = memchr((d->c.read_line.binp)->orig_bytes + 
-				  d->c.read_line.read_offset + d->c.read_line.read_size,'\n',read_size);
+				  d->c.read_line.read_offset + d->c.read_line.read_size,
+                                  delim,read_size);
 	    ASSERT(read_size <= size);
 	    d->c.read_line.read_size += read_size;
 	    if (nl_ptr != NULL) {
 		/* If found, we're done */
 		d->c.read_line.nl_pos = ((char *) nl_ptr) - 
 		    ((char *) ((d->c.read_line.binp)->orig_bytes)) + 1;
-		if (d->c.read_line.nl_pos > 1 &&
+		if (delim == '\n' && d->c.read_line.nl_pos > 1 &&
 		    *(((char *) nl_ptr) - 1) == '\r') {
 		    --d->c.read_line.nl_pos;
 		    *(((char *) nl_ptr) - 1) = '\n';
@@ -3430,13 +3436,15 @@ file_outputv(ErlDrvData e, ErlIOVec *ev) {
 	}
 	if (desc->read_size > 0) {
 	    /* look for '\n' in what we'we already got */
-	    void *nl_ptr = memchr(desc->read_binp->orig_bytes + desc->read_offset,'\n',desc->read_size);
+	    char delim   = desc->delimiter;
+	    void *nl_ptr = memchr(desc->read_binp->orig_bytes + desc->read_offset,delim,desc->read_size);
 	    if (nl_ptr != NULL) {
 		/* If found, we're done */
 		int skip = 0;
 		size_t size = ((char *) nl_ptr) - 
 		    ((char *) (desc->read_binp->orig_bytes + desc->read_offset)) + 1;
 		if (size > 1 &&
+		    delim == '\n' &&
 		    *(((char *) nl_ptr) - 1) == '\r') {
 		    *(((char *) nl_ptr) - 1) = '\n';		    
 		    skip = 1;
@@ -3474,6 +3482,7 @@ file_outputv(ErlDrvData e, ErlIOVec *ev) {
 	d->c.read_line.binp = desc->read_binp;
 	d->c.read_line.read_offset = desc->read_offset;
 	d->c.read_line.read_size = desc->read_size;
+	d->c.read_line.delimiter = desc->delimiter;
 #ifdef USE_VM_PROBES
 	dt_i1 = d->fd;
 	dt_i2 = d->flags;
@@ -4096,6 +4105,16 @@ file_outputv(ErlDrvData e, ErlIOVec *ev) {
 	    TRACE_C('K');
 	    reply_ok(desc);
 	} goto done;
+	case FILE_OPT_DELIMITER: {
+	    char delim;
+	    if (ev->size != 1+1+1
+	        || !EV_GET_CHAR(ev, &delim, &p, &q)) {
+		reply_posix_error(desc, EINVAL);
+		goto done;
+	    }
+	    desc->delimiter = delim;
+	    reply_ok(desc);
+	} goto done;
 	default:
 	    reply_posix_error(desc, EINVAL);
 	    goto done;
diff --git a/erts/preloaded/src/prim_file.erl b/erts/preloaded/src/prim_file.erl
index c87b264..8f1d0da 100644
--- a/erts/preloaded/src/prim_file.erl
+++ b/erts/preloaded/src/prim_file.erl
@@ -138,6 +138,7 @@
 %% Options
 -define(FILE_OPT_DELAYED_WRITE, 0).
 -define(FILE_OPT_READ_AHEAD,    1).
+-define(FILE_OPT_DELIMITER,     2).
 
 %% IPREAD variants
 -define(IPREAD_S32BU_P32BU, 0).
@@ -1239,6 +1240,10 @@ open_mode([{read_ahead, Size}|Rest], Mode, Portopts, Setopts)
 	true ->
 	    einval
     end;
+open_mode([{line_delimiter, Char}|Rest], Mode, Portopts, Setopts)
+  when is_integer(Char), 0 =< Char, Char =< 255 ->
+    open_mode(Rest, Mode, Portopts,
+          [<<?FILE_SETOPT, ?FILE_OPT_DELIMITER, Char>> | Setopts]);
 open_mode([], Mode, Portopts, Setopts) ->
     {Mode, reverse(Portopts), reverse(Setopts)};
 open_mode(_, _Mode, _Portopts, _Setopts) ->
diff --git a/lib/kernel/doc/src/file.xml b/lib/kernel/doc/src/file.xml
index 4954568..ebcfa86 100644
--- a/lib/kernel/doc/src/file.xml
+++ b/lib/kernel/doc/src/file.xml
@@ -767,6 +767,12 @@
               less than, or even greater than <c>Size</c> bytes, no
               performance gain can be expected.</p>
           </item>
+          <tag><c>{line_delimiter, Delimiter}</c></tag>
+          <item>
+            <p>When file is opened in the <c>raw</c> mode, <c>Delimiter</c> tells
+              <c>read_line/1</c> to use the <c>Delimiter</c> character for line
+              delimitation.  Default value for <c>Delimiter</c> is <c>$\n</c></p>
+          </item>
           <tag><c>read_ahead</c></tag>
           <item>
             <p>The same as <c>{read_ahead, Size}</c> with a reasonable
diff --git a/lib/kernel/test/file_SUITE.erl b/lib/kernel/test/file_SUITE.erl
index 8856be3..fa3857d 100644
--- a/lib/kernel/test/file_SUITE.erl
+++ b/lib/kernel/test/file_SUITE.erl
@@ -84,7 +84,7 @@
 
 -export([large_file/1, large_write/1]).
 
--export([read_line_1/1, read_line_2/1, read_line_3/1,read_line_4/1]).
+-export([read_line_1/1, read_line_2/1, read_line_3/1,read_line_4/1,read_line_5/1]).
 
 -export([advise/1]).
 
@@ -4239,6 +4239,35 @@ read_line_4(Config) when is_list(Config) ->
 	    end || {_,File,_,Y} <- All , Y =:= fail],
     ?line read_line_remove_files(All),
     ok.
+read_line_5(suite) ->
+    [];
+read_line_5(doc) ->
+    ["read_line with custom line delimiter"];
+read_line_5(Config) when is_list(Config) ->
+    PrivDir = ?config(priv_dir, Config),
+    File    = filename:join(PrivDir, "read_line_test_delim.txt"),
+    file:delete(File),
+    ok      = file:write_file(File, <<"abc",0,"efg",0,"kkk">>),
+    {ok,F1} = file:open(File, [binary,raw,{line_delimiter, 0}]),
+    {ok,<<"abc",0>>} = file:read_line(F1),
+    {ok,<<"efg",0>>} = file:read_line(F1),
+    {ok,<<"kkk">>}   = file:read_line(F1),
+    eof              = file:read_line(F1),
+    ok      = file:close(F1),
+    ok      = file:write_file(File, <<"abcXefgXkkk">>),
+    {ok,F2} = file:open(File, [binary,raw,{line_delimiter, $X}]),
+    {ok,<<"abcX">>}  = file:read_line(F2),
+    {ok,<<"efgX">>}  = file:read_line(F2),
+    {ok,<<"kkk">>}   = file:read_line(F2),
+    eof              = file:read_line(F2),
+    ok      = file:close(F2),
+    ok      = file:write_file(File, <<"abc\nefg\nkkk">>),
+    {ok,F3} = file:open(File, [binary,raw]),
+    {ok,<<"abc\n">>} = file:read_line(F3),
+    {ok,<<"efg\n">>} = file:read_line(F3),
+    {ok,<<"kkk">>}   = file:read_line(F3),
+    eof              = file:read_line(F3),
+    ok      = file:close(F3).
 
 rl_lines() ->
     [ <<"hej">>,<<"hopp">>,<<"i">>,<<"lingon\rskogen">>].
